pkg/extensions/sync/sync.go

```diff
@@ -0,0 +1,521 @@
+package sync
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Masterminds/semver"
+	"github.com/anuvu/zot/errors"
+	"github.com/anuvu/zot/pkg/log"
+	"github.com/containers/common/pkg/retry"
+	"github.com/containers/image/v5/copy"
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/signature"
+	"github.com/containers/image/v5/types"
+	"gopkg.in/resty.v1"
+)
+
+const (
+	maxRetries = 3
+	delay      = 5 * time.Minute
+)
+
+var zotCertsDir string //nolint: gochecknoglobals
+
+var locker sync.Mutex //nolint: gochecknoglobals
+
+// /v2/_catalog struct.
+type catalog struct {
+	Repositories []string `json:"repositories"`
+}
+
+type RegistryConfig struct {
+	URL          string
+	PollInterval time.Duration
+	Credentials  types.DockerAuthConfig
+	Content      []Content
+	TLSVerify    bool
+	CertDir      string
+}
+
+type Content struct {
+	Prefix string
+	Tags   *Tags
+}
+
+type Tags struct {
+	Regex  *string
+	Semver *bool
+}
+
+type PostHandler struct {
+	Address    string
+	Port       string
+	ServerCert string
+	ServerKey  string
+	CACert     string
+	Cfg        []RegistryConfig
+	Log        log.Logger
+}
+
+func (h *PostHandler) Handler(w http.ResponseWriter, r *http.Request) {
+	destCtx, policyContext, err := getDestContexts(h.ServerCert, h.ServerKey, h.CACert)
+	if err != nil {
+		WriteData(w, http.StatusInternalServerError, err.Error())
+
+		return
+	}
+
+	defer policyContext.Destroy() //nolint: errcheck
+
+	destRegistry := fmt.Sprintf("%s:%s", h.Address, h.Port)
+
+	for _, regCfg := range h.Cfg {
+		if err := syncRegistry(regCfg, h.Log, destRegistry, destCtx, policyContext); err != nil {
+			h.Log.Err(err).Msg("Error while syncing")
+			WriteData(w, http.StatusInternalServerError, err.Error())
+
+			return
+		}
+	}
+
+	WriteData(w, http.StatusOK, "")
+}
+
+func WriteData(w http.ResponseWriter, status int, msg string) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(status)
+	_, _ = w.Write([]byte(msg))
+}
+
+// getCatalog gets all repos from a registry.
+func getCatalog(regCfg RegistryConfig) (catalog, error) {
+	var c catalog
+
+	registryCatalogURL := fmt.Sprintf("%s%s", regCfg.URL, "/v2/_catalog")
+	client := resty.New()
+
+	if regCfg.TLSVerify {
+		clientCert := fmt.Sprintf("%s/client.cert", regCfg.CertDir)
+		clientKey := fmt.Sprintf("%s/client.key", regCfg.CertDir)
+		caCertPath := fmt.Sprintf("%s/ca.crt", regCfg.CertDir)
+
+		caCert, err := ioutil.ReadFile(caCertPath)
+		if err != nil {
+			return c, err
+		}
+
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+
+		client.SetTLSClientConfig(&tls.Config{RootCAs: caCertPool})
+
+		cert, err := tls.LoadX509KeyPair(clientCert, clientKey)
```

Comment:
to get a coverage for line 128, If we will pass invalid key pair e.g (different public key corresponding to private key) it should give error.